### PR TITLE
Smoke-test gate for technical PRs (closes #14)

### LINKED ISSUE #14: Smoke test for technical changes
Approval should only be given when smoke tested
https://github.com/Panenco/argenx-argo-map/pull/270 for instance, major change but wasn't tested

### PR DESCRIPTION
**Pull request by ljobse**
## Summary

Closes #14. The reviewer was happy to APPROVE PRs whose stated intent is "no user-visible behavior change" — refactors, library swaps, framework/runtime upgrades, build/config changes — without ever running the app. Such PRs have no acceptance criteria for the spec-validation path to bind to ("nothing should change for the user"), so the only regression-catching path is to actually walk through a user flow. The cited example is [argenx-argo-map#270](https://github.com/Panenco/argenx-argo-map/pull/270) — a major upgrade that merged unverified.

This adds a single-judgement-boolean APPROVE gate, modeled byte-for-byte on PR #15's `manual_spec_present` gate. The two compose naturally:

> APPROVE is only granted when *something* substantively validated the change — either a human-authored spec, or a working app.

## How it works

1. **Test planner** detects "no behavior change" intent from PR title/body keywords (`refactor`, `chore`, `bump`, `upgrade`, `migrate`), diff shape (high move/rename ratio, low net new logic), and dependency-manifest deltas across **any ecosystem** (npm, poetry/uv/pip, go, cargo, gem, maven/gradle, `Dockerfile`, `.tool-versions`, GitHub Actions majors). When triggered, it emits `## Technical change: true` in `test-plan.md` and selects strategy `functional`.
2. **Functional tester** picks one representative user flow autonomously based on which code paths the change affects (refactor → flow that drives the refactored module's public surface; library swap → flow with formatting/transforms that used the old library; framework upgrade → route using framework features), walks through it with screenshots, axe-core a11y audit, console-error check.
3. **Verdict gate** (`build-review.sh`) downgrades APPROVE → COMMENT when `Technical change: true` AND the smoke run did not return PASS/WARN — including the degraded-mode case where `dev-start.sh` is missing or the tester crashed. A banner in the review body explains how to enable APPROVE.

## Why a single judgement boolean

Same reasoning as PR #15: an LLM-judged boolean is more robust than a regex on conventional commit prefixes. New refactor patterns and ecosystems slip through brittle string matching; LLM judgement on PR-body prose ("no behavior change", "pure refactor", "behavior-preserving") catches them. Folding the boolean into the file the planner **already** writes (`test-plan.md`) — and reading it directly in `build-review.sh` rather than hopping through the tester's `functional-meta.json` — also makes the flag **survive tester crashes**, which is exactly the case the gate needs to catch.

## Crash robustness

The workflow injects a synthetic `{strategy: "skip", overall: "PASS"}` into `functional-meta.json` when the tester exits non-zero, so downstream `jq` never sees a missing file. Without care, this would silently bypass the new gate (the placeholder makes `FUNCTIONAL_OVERALL == "PASS"` even though no smoke run happened). Two mitigations:

- `TECHNICAL_CHANGE` is read from `test-plan.md`, not `functional-meta.json` — the planner always completes before the tester runs, so the flag survives.
- `SMOKE_OK` checks `FUNCTIONAL_OK` (the workflow-set crash indicator) in addition to `FUNCTIONAL_OVERALL`, so a synthetic-on-crash placeholder still triggers the gate.

## Files changed

| File | Change |
|---|---|
| `skills/review-test-planner.md` | Classification table 3→5 rows, technical-change detection signals (PR title/body/diff/linked-issue), smoke-scenario template (rewrites the empty `quick`-strategy subsection), documents `## Technical change: true` front-matter flag. |
| `skills/review-functional-tester.md` | One bullet under Output: copy `Technical change: true` from `test-plan.md` into `functional-meta.json` for the artifact record. |
| `scripts/build-review.sh` | Read `TECHNICAL_CHANGE` from `test-plan.md`. Derive `SMOKE_OK` from `FUNCTIONAL_OK + FUNCTIONAL_OVERALL`. New verdict-ladder branch (after `manual_spec_present`). New banner. Propagate `technical_change` to `review-result.json`. |
| `scripts/verdict-gate.sh` | Mirror banner in GitHub Actions step summary. |
| `README.md` | Short "Smoke-test gate for technical PRs" paragraph next to the existing spec-presence paragraph. |

Net change: +108 / −10 across 5 files. No new files, no new pipeline contracts.

## Test plan

- [ ] Open this PR — it links #14 (a real human-written GitHub issue) and is mostly review-pipeline plumbing (planner classifies as `skip`); verdict should be APPROVE.
- [ ] **Technical-PR happy path** — follow-up PR that bumps a major dep in a consumer repo with a working `.github/claude-review/dev-start.sh`. Expect: `test-plan.md` has `Strategy: functional` + `Technical change: true`, smoke runs and posts screenshots inline, `functional_meta.overall == "PASS"`, verdict APPROVE.
- [ ] **Technical-PR degraded mode** — dep-bump PR in a repo *without* `dev-start.sh`. Expect: `Technical change: true`, functional tester does not run, verdict COMMENT with the new banner visible in both the review body and the step summary.
- [ ] **Refactor (no version bump) trigger** — PR titled `refactor(...)` that moves/renames code without a manifest touch. Expect: planner classifies as `Technical change: true` from the title + diff-shape signal alone.
- [ ] **Cross-ecosystem trigger** — PR touching only a non-Node manifest (`Cargo.toml`, `Dockerfile FROM`, `pyproject.toml`). Expect: still classified as `Technical change: true`.
- [ ] Spot-check `review-result.json` in workflow artifacts to confirm the new `technical_change` field is present.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes the pipeline’s verdict logic and can newly withhold `APPROVE` on certain PRs based on smoke-test status, so misclassification or edge cases in functional-meta handling could affect merge gating.
> 
> **Overview**
> Adds a new **smoke-test approval gate** for PRs whose intent is “no user-visible behavior change”: the test planner can emit `## Technical change: true` in `test-plan.md`, and `build-review.sh` now downgrades `APPROVE` to `COMMENT` unless functional validation actually ran and returned `PASS`/`WARN` (including crash/degraded-mode cases where the tester didn’t run).
> 
> Propagates the new `technical_change`/`smoke_ok` signals into `review-result.json` and surfaces explanatory banners in both the posted PR review body and the Actions step summary (`verdict-gate.sh`). Updates planner/tester skill docs and README to define the flag, detection heuristics, and the “one end-to-end flow” smoke scenario for technical changes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit daa362866fb21826bca78ad72eeaf1d9c91a03f0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: scripts/build-review.sh [hunk 1]
```diff
@@ -306,6 +306,39 @@ HUMAN_REVIEW=$(echo "$CORE_META" | jq -r '.requires_human_review // false')
...
+
+# Did the smoke test actually pass? Account for the workflow's synthetic
+# `{strategy:"skip",overall:"PASS"}` placeholder injected on tester crash —
+# without this guard, FUNCTIONAL_OVERALL would read "PASS" for a crashed
+# tester and the technical-change gate would be silently bypassed in
+# exactly the case it's meant to catch.
+SMOKE_OK=false
+if [ "${FUNCTIONAL_OK:-1}" -eq 1 ] && { [ "$FUNCTIONAL_OVERALL" = "PASS" ] || [ "$FUNCTIONAL_OVERALL" = "WARN" ]; }; then
+  SMOKE_OK=true
+fi
```
[the remark this answers is not in the document]
Comment (ljobse) [reply scripts/build-review.sh:353]: Fixed in commit 21933c7. Restructured SMOKE_OK as early-guards and added a `FUNCTIONAL_STRATEGY != "skip"` check — the synthetic placeholder's strategy=skip now correctly disqualifies SMOKE_OK in the degraded-mode case you flagged. Verified across all six failure/success modes.

File: scripts/verdict-gate.sh [hunk 2]
```diff
@@ -76,6 +78,10 @@ MANUAL_SPEC=$(jq -r 'if (type == "object" and has("manual_spec_present")) then .
     echo ""
     echo "> :no_entry: **No manual spec available — APPROVE withheld.** Link an issue, paste acceptance criteria, or wire up an external tracker to enable APPROVE."
   fi
+  if [ "$TECHNICAL_CHANGE" = "true" ] && [ "$FUNCTIONAL_OVERALL" != "PASS" ] && [ "$FUNCTIONAL_OVERALL" != "WARN" ]; then
+    echo ""
+    echo "> :no_entry: **Technical change — APPROVE withheld until smoke-tested** (overall=\`$FUNCTIONAL_OVERALL\`). Refactors/upgrades have no acceptance criteria, so a passing smoke run is required. Configure \`.github/claude-review/dev-start.sh\` to bring up the app, or fix the issues that caused the smoke run to fail."
+  fi
```
[the remark this answers is not in the document]
Comment (ljobse) [reply scripts/verdict-gate.sh:93]: Fixed in commit d014878. Persisted `smoke_ok` to review-result.json so verdict-gate.sh evaluates the same condition as build-review.sh. Subsequent commit 21933c7 also tightened smoke_ok itself to reject the degraded-mode synthetic.

File: scripts/build-review.sh [hunk 3]
```diff
@@ -306,6 +306,39 @@ HUMAN_REVIEW=$(echo "$CORE_META" | jq -r '.requires_human_review // false')
...
+  TECHNICAL_CHANGE=true
+fi
+
+# Did the smoke test actually pass? Account for the workflow's synthetic
+# `{strategy:"skip",overall:"PASS"}` placeholder injected on tester crash —
+# without this guard, FUNCTIONAL_OVERALL would read "PASS" for a crashed
+# tester and the technical-change gate would be silently bypassed in
+# exactly the case it's meant to catch.
+SMOKE_OK=false
+if [ "${FUNCTIONAL_OK:-1}" -eq 1 ] && { [ "$FUNCTIONAL_OVERALL" = "PASS" ] || [ "$FUNCTIONAL_OVERALL" = "WARN" ]; }; then
```
[the remark this answers is not in the document]
Comment (ljobse) [reply scripts/build-review.sh]: Fixed in commit 21933c7. SMOKE_OK is now an early-guard ladder (FUNCTIONAL_OK → STRATEGY → OVERALL) — same approach you suggested. Also closes a separate bug Cursor caught about the synthetic placeholder bypassing the gate.

File: scripts/verdict-gate.sh [hunk 4]
```diff
@@ -76,6 +78,10 @@ MANUAL_SPEC=$(jq -r 'if (type == "object" and has("manual_spec_present")) then .
     echo ""
     echo "> :no_entry: **No manual spec available — APPROVE withheld.** Link an issue, paste acceptance criteria, or wire up an external tracker to enable APPROVE."
   fi
+  if [ "$TECHNICAL_CHANGE" = "true" ] && [ "$FUNCTIONAL_OVERALL" != "PASS" ] && [ "$FUNCTIONAL_OVERALL" != "WARN" ]; then
+    echo ""
+    echo "> :no_entry: **Technical change — APPROVE withheld until smoke-tested** (overall=\`$FUNCTIONAL_OVERALL\`). Refactors/upgrades have no acceptance criteria, so a passing smoke run is required. Configure \`.github/claude-review/dev-start.sh\` to bring up the app, or fix the issues that caused the smoke run to fail."
```
[the remark this answers is not in the document]
Comment (ljobse) [reply scripts/verdict-gate.sh]: The condition collapsed to two checks (`TECHNICAL_CHANGE = true && SMOKE_OK = false`) after commit d014878 moved the FUNCTIONAL_OVERALL evaluation into smoke_ok. The remaining 2-clause guard is short enough to leave inline; happy to split further if you'd still prefer.

File: skills/review-functional-tester.md [hunk 5]
```diff
@@ -158,6 +159,8 @@ Always write both files, even on partial completion.
 }
 ```
 
+- `technical_change` — copy verbatim from `test-plan.md`'s front matter. If the planner emitted `## Technical change: true`, set `true`. Otherwise set `false` (or omit). The verdict gate uses this to withhold APPROVE when a refactor/upgrade PR was not actually smoke-tested — including the degraded-mode case where the dev environment couldn't be brought up at all.
+
```
[the remark this answers is not in the document]
Comment (ljobse) [reply skills/review-functional-tester.md:163]: Fixed in commit daa3628. Applied your suggested wording verbatim — accurate on both 'heading not front matter' and 'test-plan.md is the source of truth, functional-meta.json is artifact-only'.
